File: src/main/ruby/advent2023/Puzzle2.rb
load 'base.rb'

class Puzzle2
  include Base
  
  class Game
    attr_reader :game
    attr_reader :turns

    class Turn
      attr_reader :parts

      def initialize(str)
        @parts = str.split(",").map do |s| 
          s = s.strip
          num = s.to_i
          color = s[s.index(" ")..].strip.to_sym
          [color, num]
        end.to_h
      end

      def has_enough_blocks(blocks)
        blocks.all? do |color, num|
          @parts.fetch(color, 0) <= num
        end
      end
    end
  
    def initialize(str)
      str = str = str["Game ".length..]
      @game = str[0..str.index(":")-1].to_i    
  
      str = str[str.index(":")+2..]
      @turns = str.split(";").map { |s| s.strip }.map { |s| Turn.new s }
    end

    def has_enough_blocks(blocks)
      @turns.all? { |b| b.has_enough_blocks(blocks) }
    end

    def min_possible_blocks
      ret = {:red=>0, :blue=>0, :green=>0}
      @turns.each do |t|
        t.parts.each do |key, value|
          ret[key] = value if value > ret[key]
        end
      end
      ret
    end
  end

  def main
    solution1 = self.solve_puzzle("puzzle2/inputs.txt", Array.new) do |line, data|
        data << parse_game(line)
    end
    puts "Solution1: #{solution1}"

    solution2 = self.solve_puzzle2("puzzle2/inputs.txt", Array.new) do |line, data|
        data << parse_game(line)
    end
    puts "Solution2: #{solution2}"
  end

  def compute_solution(data)
    ok_games = data.filter {|g| g.has_enough_blocks(:red=>12, :green=>13, :blue=>14)}
    ok_games.sum { |g| g.game }
  end

  def compute_solution2(data)
    mins = data.map do |g|
      g.min_possible_blocks
    end
    mins.sum do |m|
      m.values.reduce(1, :*)
    end
  end

  def parse_game(str)
    Game.new str
  end
end

p = Puzzle2.new
p.main

File: src/main/python/advent2024/Puzzle2.py
from Base import Base

Report = list[int]
MyDataType = list[Report]

class Puzzle(Base[MyDataType, int, int]):
    def main(self):
      solution1: int = self.solve_puzzle("puzzle2/inputs.txt", [])  # type: ignore
      print("Solution1: {}".format(solution1))

      # 417 too high
      solution2: int = self.solve_puzzle2("puzzle2/inputs.txt", []) # type: ignore
      print("Solution2: {}".format(solution2))

    def read_one_line(self, line: str, data: MyDataType):
      report = list(map(int, line.split()))
      data.append(report)

    def compute_solution(self, data: MyDataType) -> int:
      safe = 0
      for report in data:
          if self.report_is_safe(report):
              safe += 1
      return safe

    def compute_solution2(self, data: MyDataType):
      safe = 0
      for report in data:
          bad_index = self.report_bad_index(report)
          if bad_index < 0 or self.safe_if_remove_around(bad_index, report):
              safe += 1
      return safe

    def report_bad_index(self, report: Report) -> int:
        increasing = report[0] < report[1]
        i = 0
        while i < len(report) - 1:
            if not self.neighbors_are_safe(report[i], report[i + 1], increasing):
                return i
            i += 1
        return -1

    def report_is_safe(self, report: Report) -> int:
        return self.report_bad_index(report) < 0

    def safe_if_remove_around(self, bad_index: int, report: Report) -> bool:
        ret = False
        if bad_index > 0:
            ret = self.report_is_safe(self.remove_element(bad_index - 1, report))
        ret = ret or self.report_is_safe(self.remove_element(bad_index, report))
        ret = ret or self.report_is_safe(self.remove_element(bad_index + 1, report))
        return ret

    def remove_element(self, index: int, report: Report) -> Report:
        return report[:index] + report[index+1:]

    @staticmethod
    def neighbors_are_safe(first: int, after: int, increasing: bool) -> bool:
        this_increasing = first < after
        diff = abs(after - first)
        return not (this_increasing != increasing or diff == 0 or diff > 3)

puzzle: Puzzle = Puzzle()
puzzle.main()